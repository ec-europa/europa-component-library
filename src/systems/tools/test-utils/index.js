const merge = require('deepmerge');
const twing = require('@ecl-twig/ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const test = document.createElement('test');
      test.innerHTML = html.trim();
      resolve(test);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
