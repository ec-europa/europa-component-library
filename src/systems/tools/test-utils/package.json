{
  "name": "@ecl-twig/test-utils",
  "version": "2.31.1",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
