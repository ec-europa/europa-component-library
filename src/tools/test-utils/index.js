const merge = require('deepmerge');
const twing = require('@ecl/twig-ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const div = document.createElement('div');
      div.innerHTML = html.trim();
      resolve(div.firstChild);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
