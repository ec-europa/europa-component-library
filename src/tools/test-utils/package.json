{
  "name": "@ecl/test-utils",
  "version": "3.4.2",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
