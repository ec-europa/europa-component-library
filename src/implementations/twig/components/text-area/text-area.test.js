import { merge, renderTwigFileAsNode } from '@ecl/test-utils';
import specDefault from '@ecl/specs-component-text-area/demo/data';

const specInvalid = { ...specDefault, invalid: true };
const specDisabled = { ...specDefault, disabled: true };

const testTextArea = (dataDefault, dataInvalid, dataDisabled) => {
  const template = '@ecl/text-area/text-area.html.twig';
  const render = (params) => renderTwigFileAsNode(template, params);

  describe('Optional', () => {
    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(dataDefault)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra group class names', () => {
      expect.assertions(1);

      const optionsWithExtraGroupClasses = merge(dataDefault, {
        extra_group_classes: 'custom-group-class custom-group-class--test',
      });

      return expect(
        render(optionsWithExtraGroupClasses)
      ).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(dataDefault, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra label class names', () => {
      expect.assertions(1);

      const optionsWithExtraLabelClasses = merge(dataDefault, {
        extra_label_classes: 'custom-label-class custom-label-class--test',
      });

      return expect(
        render(optionsWithExtraLabelClasses)
      ).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(dataDefault, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('Required', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      const optionsWhenRequired = merge(dataDefault, {
        required: true,
      });

      return expect(render(optionsWhenRequired)).resolves.toMatchSnapshot();
    });
  });

  describe('Disabled', () => {
    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(dataDisabled)).resolves.toMatchSnapshot();
    });
  });

  describe('Invalid', () => {
    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(dataInvalid)).resolves.toMatchSnapshot();
    });
  });
};

describe('Text area', () => {
  testTextArea(specDefault, specInvalid, specDisabled);
});
