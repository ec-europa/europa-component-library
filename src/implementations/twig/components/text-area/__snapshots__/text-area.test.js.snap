// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text area Default renders correctly 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Default renders correctly when required 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Default renders correctly with a default value 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  >
    Hello world!
  </textarea>
</div>
`;

exports[`Text area Default renders correctly with extra attributes 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Default renders correctly with extra class names 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m custom-class custom-class--test"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Default renders correctly with extra group class names 1`] = `
<div
  class="ecl-form-group custom-group-class custom-group-class--test"
>
  <label
    class="ecl-form-label"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Default renders correctly with extra label class names 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label custom-label-class custom-label-class--test"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Disabled renders correctly 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--disabled"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block ecl-help-block--disabled"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    disabled=""
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Disabled renders correctly with extra attributes 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--disabled"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block ecl-help-block--disabled"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
    disabled=""
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Disabled renders correctly with extra class names 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--disabled"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block ecl-help-block--disabled"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m custom-class custom-class--test"
    disabled=""
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Disabled renders correctly with extra group class names 1`] = `
<div
  class="ecl-form-group custom-group-class custom-group-class--test"
>
  <label
    class="ecl-form-label ecl-form-label--disabled"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block ecl-help-block--disabled"
  >
    This is the input's helper text.
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--m"
    disabled=""
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Invalid renders correctly 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <div
    class="ecl-feedback-message"
  >
    This is the error message
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--invalid ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Invalid renders correctly with extra attributes 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <div
    class="ecl-feedback-message"
  >
    This is the error message
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--invalid ecl-text-area--m"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Invalid renders correctly with extra class names 1`] = `
<div
  class="ecl-form-group"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <div
    class="ecl-feedback-message"
  >
    This is the error message
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--invalid ecl-text-area--m custom-class custom-class--test"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;

exports[`Text area Invalid renders correctly with extra group class names 1`] = `
<div
  class="ecl-form-group custom-group-class custom-group-class--test"
>
  <label
    class="ecl-form-label ecl-form-label--invalid"
    for="example-textarea-id-1"
  >
    Label
    <span
      class="ecl-form-label__required"
    >
      *
    </span>
  </label>
  <div
    class="ecl-help-block"
  >
    This is the input's helper text.
  </div>
  <div
    class="ecl-feedback-message"
  >
    This is the error message
  </div>
  <textarea
    class="ecl-text-area ecl-text-area--invalid ecl-text-area--m"
    id="example-textarea-id-1"
    required=""
    rows="4"
  />
</div>
`;
